Issue #505: Addon module redo (#563)
plone/plone.api: docs/addon.md

---
myst:
  html_meta:
    "description": "Get and manage add-ons"
    "property=og:description": "Get, modify and manage add-ons"
    "property=og:title": "add-ons"
    "keywords": "Plone, API, development"
---

```{eval-rst}
.. module:: plone.api.addon
```

(chapter-addons)=

# Add-on

(addons-get-addons)=

## Get add-ons

To get all the add-ons present in the current Plone site, use the {func}`api.addon.get_addons` function.
The function accepts an optional `limit` parameter to filter the returned add-ons.

```python
from plone import api

# Get all add-ons
addons = api.addon.get_addons()

# Get only installed add-ons

> installed = api.addon.get_addons(limit='installed')

Reviewer: ```suggestion
installed = api.addon.get_addons(limit="installed")
```

# Get only upgradable add-ons

> upgradable = api.addon.get_addons(limit='upgradable')

Reviewer: ```suggestion
upgradable = api.addon.get_addons(limit="upgradable")
```

# Get only broken add-ons

> broken = api.addon.get_addons(limit='broken')

Reviewer: ```suggestion
broken = api.addon.get_addons(limit="broken")
```

# Get IDs of installed add-ons
addon_ids = api.addon.get_addon_ids(limit='installed')
```

The `limit` parameter accepts these values:
- `installed`: Return only installed add-ons.
- `upgradable`: Return only add-ons that can be upgraded.
- `broken`: Return broken add-ons.
- `non_installable`: Return non-installable add-ons.

## Get add-on information

> To get information about a specific add-on.

Reviewer: ```suggestion
To get information about a specific add-on, use the {func}`api.addon.get` function, passing in the name of the add-on as a string.
```

```python
from plone import api

> addon = api.addon.get('plone.session')

Reviewer: ```suggestion
addon = api.addon.get("plone.session")
```

print(addon.id)           # ID of the add-on
print(addon.version)      # Version string
print(addon.title)        # Display title
print(addon.description)  # Description

> print(addon.flags)        # List of flags like ['installed', 'upgradable']

Reviewer: ```suggestion
print(addon.flags)        # List of flags like ["installed", "upgradable"]
```

```

## Install and uninstall add-ons

> To install an add-on:

Reviewer: ```suggestion
To install an add-on, use the {func}`api.addon.install` function, passing in the name of the add-on as a string, as shown in the following example.
```

```python
from plone import api

> success = api.addon.install('plone.session')

Reviewer: ```suggestion
success = api.addon.install("plone.session")
```

```

> To uninstall an add-on:

Reviewer: ```suggestion
To uninstall an add-on, use the {func}`api.addon.uninstall` function, passing in the name of the add-on as a string, as shown in the following example.

```

```python
from plone import api

> success = api.addon.uninstall('plone.session')

Reviewer: ```suggestion
success = api.addon.uninstall("plone.session")
```

```

## Get add-on version

> To get the version of an add-on.

Reviewer: ```suggestion
To get the version of an add-on, use the {func}`api.addon.get_version` function, passing in the name of the add-on as a string, as shown in the following example.
```

```python
from plone import api

> version = api.addon.get_version('plone.session')

Reviewer: ```suggestion
version = api.addon.get_version("plone.session")
```

```

(addons-exceptions)=

## Exceptions

> The following exceptions may be raised.
> 
> - {exc}`~plone.api.exc.InvalidParameterError` exception is raised in the following cases:
>   - Trying to get information about a non-existent add-on
>   - Using an invalid limit parameter value

Reviewer: ```suggestion

The {exc}`~plone.api.exc.InvalidParameterError` exception may be raised in the following cases.

- When using the {func}`api.addon.get` function, trying to get information about a non-existent add-on.
- When using either function {func}`api.addon.get_addons` or {func}`api.addon.get_addon_ids`, using an invalid `limit` parameter value.

```

> 
> ## Further reading
> 
> For more information on possible flags and usage options please see the full {ref}`plone-api-addon` specification.

Reviewer: "If it ain't documented, it's broken." Avoid telling users to read code as documentation.

Additionally, each method gets automatically linked to its source code, so this is redundant.

```suggestion
```

---

```{eval-rst}
.. module:: plone.api.addon
```

(chapter-addons)=

# Add-ons

This chapter describes how to get, update, install, uninstall, and manage Plone add-ons.

(addons-get-addons)=

## Get add-ons

To get all the add-ons present in the current Plone site, use the {func}`api.addon.get_addons` function.
The function accepts an optional `limit` parameter to filter the returned add-ons.
`limit` may be one of the following strings.

`available`
:   Products that are not installed, but could be.

`broken`
:   Uninstallable products with broken dependencies.

`installed`
:   Only products that are installed and not hidden.

`non_installable`
:   Non-installable products.

`upgradable`
:   Only products with upgrades.

The following examples demonstrate usage of the {func}`api.addon.get_addons` function.

```python
from plone import api

# Get all add-ons
addons = api.addon.get_addons()

# Get only installed add-ons
installed = api.addon.get_addons(limit="installed")

# Get only upgradable add-ons
upgradable = api.addon.get_addons(limit="upgradable")

# Get only broken add-ons
broken = api.addon.get_addons(limit="broken")
```

(addons-get-addon-ids)=

## Get add-on IDs

> To get all the add-ons' IDs present in the current Plone site, use the {func}`api.addon.get_addon_ids` function.

Reviewer 2: ```suggestion
To get the IDs of all the add-ons present in the current Plone site, use the {func}`api.addon.get_addon_ids` function.
```

The following examples demonstrates usage of the {func}`api.addon.get_addon_ids` function.

```python
# Get IDs of installed add-ons
addon_ids = api.addon.get_addon_ids(limit="installed")
```
## Get add-on information

To get information about a specific add-on, use the {func}`api.addon.get` function, passing in the name of the add-on as a string.

```python
from plone import api

addon = api.addon.get("plone.session")
print(addon.id)           # ID of the add-on
print(addon.version)      # Version string
print(addon.title)        # Display title
print(addon.description)  # Description
print(addon.flags)        # List of flags like ["installed", "upgradable"]
```

## Install and uninstall add-ons

To install an add-on, use the {func}`api.addon.install` function, passing in the name of the add-on as a string, as shown in the following example.

```python
from plone import api

success = api.addon.install("plone.session")
```

To uninstall an add-on, use the {func}`api.addon.uninstall` function, passing in the name of the add-on as a string, as shown in the following example.

```python
from plone import api

success = api.addon.uninstall("plone.session")

> ```

Reviewer 2: We should mention what happens if the add-on is already installed or already uninstalled.

## Get add-on version

To get the version of an add-on, use the {func}`api.addon.get_version` function, passing in the name of the add-on as a string, as shown in the following example.

```python
from plone import api

version = api.addon.get_version("plone.session")

> ```

Reviewer 2: We should clarify that this gives the version of the Python package that was installed from PyPI, not the version of the add-on's GenericSetup profile.

```

## Install and uninstall add-ons

To install an add-on, use the {func}`api.addon.install` function, passing in the name of the add-on as a string, as shown in the following example.

```python
from plone import api

success = api.addon.install("plone.session")
```
This function returns a `false` boolean value in the following cases.
- The installation fails due to an error.
- The add-on is already installed.
- The add-on is not found among the available add-ons.

To uninstall an add-on, use the {func}`api.addon.uninstall` function, passing in the name of the add-on as a string, as shown in the following example.

```python
from plone import api

success = api.addon.uninstall("plone.session")
```

This function returns a `false` boolean value in the following cases.
- The removal of add-on fails due to an error.
- The add-on is not installed.

## Get add-on version

To get the version of an add-on, use the {func}`api.addon.get_version` function, passing in the name of the add-on as a string, as shown in the following example.

```python
from plone import api

version = api.addon.get_version("plone.session")
```

> Note that this returns the version of the Python package installed from **PyPi**, not the version of the add-on's **_GenericSetup_** profile.

Reviewer 2: ```suggestion
Note that this returns the version of the Python package installed from **PyPI**, not the version of the add-on's **_GenericSetup_** profile.
```

(chapter-addons)=

# Add-ons

This chapter describes how to get, update, install, uninstall, and manage Plone add-ons.

(addons-get-addons)=

## Get add-ons

To get all the add-ons present in the current Plone site, use the {func}`api.addon.get_addons` function.
The function accepts an optional `limit` parameter to filter the returned add-ons.
`limit` may be one of the following strings.

`available`
:   Products that are not installed, but could be.

`broken`
:   Uninstallable products with broken dependencies.

`installed`
:   Only products that are installed and not hidden.

`non_installable`
:   Non-installable products.

`upgradable`
:   Only products with upgrades.

> The following examples demonstrate usage of the {func}`api.addon.get_addons` function.

Reviewer: Whoopsie! This needs a blank line, else it becomes a definition list parsing error.

```suggestion

The following examples demonstrate usage of the {func}`api.addon.get_addons` function.
```
